File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/OverlaySource.java
package ch.bailu.aat_lib.dispatcher;

import javax.annotation.Nonnull;

import ch.bailu.aat_lib.app.AppContext;
import ch.bailu.aat_lib.preferences.OnPreferencesChanged;
import ch.bailu.aat_lib.preferences.StorageInterface;
import ch.bailu.aat_lib.preferences.map.SolidOverlayFile;

public class OverlaySource extends FileSource {
    private final SolidOverlayFile soverlay;
    private boolean enabled = true;

    private final OnPreferencesChanged onPreferencesChanged = new OnPreferencesChanged() {
        @Override
        public void onPreferencesChanged(@Nonnull StorageInterface s, @Nonnull String key) {
            if (soverlay.hasKey(key)) {
                initAndUpdateOverlay();
            }
        }
    };

    public OverlaySource(AppContext context, int iid) {
        super(context, iid);
        soverlay = new SolidOverlayFile(context.getStorage(), context, iid);
        initAndUpdateOverlay();
    }

    public void initAndUpdateOverlay() {
        setFileID(soverlay.getValueAsString());

        if (enabled && soverlay.isEnabled()) {
            super.enable();
        } else {
            super.disable();
        }
    }

    @Override
    public void enable() {
        enabled = true;
        if (soverlay.isEnabled()) {
            super.enable();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        soverlay.unregister(onPreferencesChanged);
    }

    @Override
    public void onResume() {
        super.onResume();
        soverlay.register(onPreferencesChanged);
        initAndUpdateOverlay();
    }

    @Override
    public void disable() {
        enabled = false;
        super.disable();
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/OverlaySource.kt
package ch.bailu.aat_lib.dispatcher

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.gpx.InfoID
import ch.bailu.aat_lib.preferences.OnPreferencesChanged
import ch.bailu.aat_lib.preferences.map.SolidCustomOverlay
import ch.bailu.aat_lib.preferences.map.SolidOverlayInterface
import ch.bailu.aat_lib.preferences.map.SolidPoiOverlay

class OverlaySource private constructor(
    context: AppContext,
    private val soverlay: SolidOverlayInterface
) : FileSource(context, soverlay.getIID()) {
    private val onPreferencesChanged = OnPreferencesChanged { _, key ->
        if (soverlay.hasKey(key)) {
            initAndUpdateOverlay()
        }
    }

    init {
        initAndUpdateOverlay()
    }

    private fun initAndUpdateOverlay() {
        setFile(soverlay.getValueAsFile())
        super.setEnabled(soverlay.isEnabled())
    }

    override fun setEnabled(enabled: Boolean) {
        soverlay.setEnabled(enabled)
        super.setEnabled(soverlay.isEnabled())
    }

    override fun onPause() {
        super.onPause()
        soverlay.unregister(onPreferencesChanged)
    }

    override fun onResume() {
        super.onResume()
        soverlay.register(onPreferencesChanged)
        initAndUpdateOverlay()
    }

    companion object {
        fun factoryPoiOverlaySource(context: AppContext): OverlaySource {
            return OverlaySource(context, SolidPoiOverlay(context.dataDirectory))
        }

        fun factoryCustomOverlaySource(context: AppContext, index: Int): OverlaySource {
            return OverlaySource(
                context,
                SolidCustomOverlay(context.storage, context, InfoID.OVERLAY + index)
            )
        }
    }
}
